logstash/class-logger.php

```diff
@@ -0,0 +1,426 @@
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_id'] ) && ! is_int( $data['user_id'] ) || $data['user_id'] < 0 ) {
+			trigger_error( 'Invalid `user_id` in call to ' . esc_html( __METHOD__ ) . '(). Must be an integer >= 0.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_ua'] ) && ! is_string( $data['user_ua'] ) ) {
+			trigger_error( 'Invalid `user_ua` in call to ' . esc_html( __METHOD__ ) . '(). Must be a string.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_ua'] ) && strlen( $data['user_ua'] ) > static::MAX_ENTRY_USER_UA_SIZE ) {
+			trigger_error( 'Invalid `user_ua` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_USER_UA_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['extra'] ) && ! is_array( $data['extra'] ) && ! is_object( $data['extra'] ) && ! is_scalar( $data['extra'] ) ) {
+			trigger_error( 'Invalid `extra` in call to ' . esc_html( __METHOD__ ) . '(). Must be an ojbect, array, or scalar value.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['extra'] ) && strlen( wp_json_encode( $data['extra'] ) ) > static::MAX_ENTRY_EXTRA_SIZE ) {
+			trigger_error( 'Invalid `extra` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_EXTRA_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+		}
+
+		// Adjust `feature` and JSON-encode pretty-print data.
+		$data['feature']        = 'a8c_vip_' . $data['feature'];
+
+		// Adjust `extra` and JSON-encode data that gets sent in the API request.
+		$data['extra'] = isset( $data['extra'] ) ? wp_json_encode( $data['extra'], JSON_PRETTY_PRINT ) : '';
+
+		// Adds a new log entry.
+		static::$entries[] = $data;
+
+		// Sends data to logstash on shutdown.
+		if ( ! has_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] ) ) {
+			add_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] );
+		}
+	}
+
+	/**
+	 * Sends data to logstash via REST API on shutdown, viewable in Kibana.
+	 *
+	 * @since 2020-01-10
+	 */
+	public static function process_entries_on_shutdown() : void {
+		if ( static::$processed_entries ) {
+			return; // Already done.
+		}
+		static::$processed_entries = true;
+		$endpoint = 'https://public-api.wordpress.com/rest/v1.1/logstash/bulk';
+
+		$entry_chunks = array_chunk( static::$entries, static::BULK_ENTRIES_COUNT );
+		// Process all entries.
+		foreach ( $entry_chunks as $entries ) {
+			if ( ! defined( 'VIP_GO_ENV' ) || ! VIP_GO_ENV ) {
+				static::maybe_wp_debug_log_entries( $entries );
+				continue; // Bypassing REST API below in this case.
+			}
+			$json_data     = wp_json_encode( $entries );
+			// Send to logstash via REST API endpoint with payload containing log entry details.
+			$_wp_remote_response         = wp_remote_post(
```

Comment:
Let's switch this to use `vip_safe_wp_remote_request()` instead.

Reply:
Changed in 24fbbbb92f0bb45b52fcaf26893a67b6e0952e3d